### DISCUSSION TITLE
prevent deleting the node using deleteKeyCode for specific Nodes

### DISCUSSION THREAD

**Discussion by alikhaniYepco** [xyflow/xyflow#2363] [Q&A] [answered]
How can I prevent removing the node with the property ` indelible: True ` , when the user presses the key delete or backspace
My ReactFlow Props is `deleteKeyCode: ['Backspace', 'Delete']`

node Data 
```
data: {
	indelible: true,
	label: 'End',
	source: 0,
	target: 1,
	style: {
		width: 50,
		height: 50,
		border-radius: '50%',
		font-size: '0.8rem',
	},
}
```

**Comment by moklick** [comment, marked as answer]
I think we will add a `deletable` flag for the nodes in v11. It's way easier to handle on in lib land than in user land. However you could achieve it by creating your own `onNodesChange` function like this:

```js
const onNodesChange = useCallback((changes) => 
  setNodes((nds) => {
    const parsedChanges = changes.reduce((res, change) => {
      const validChange = change.type !== 'remove' || (change.type === 'remove' && nodes.find(n => n.id === change.id)?.data.deletable);
      if (validChange) {
        res.push(change)
      }
    }, []);

    return applyNodeChanges(parsedChanges, nds));
  },
    [setNodes]
  );
```

With this solution you would need to set `data.deletable: true` for a node in order to be able to delete it.

**Reply by navidkhalilian** [reply]
Hi dude.
I think the code need a bit change to correct work :

`
const onNodesChange = useCallback((changes) => 
  setNodes((nds) => {
    const parsedChanges = changes.reduce((res, change) => {
      const validChange = change.type !== 'remove' || (change.type === 'remove' && nds.find(n => n.id === change.id)?.data.deletable);
      if (validChange) {
        res.push(change)
      }
    }, []);
    return applyNodeChanges(parsedChanges, nds));
  },
    [setNodes]
  );
`

Thanks to save life.
@moklick

**Comment by hamidsaifi77** [comment]
After searching extensively, I found a solution as follows:

```
 const onNodesChange = React.useCallback(
        (changes) => {
            const id = changes[0]?.id // this is the node id you dont want to delete 
            if (id === "1") {
                return
            }
            FLOW.setNodes((nds) => applyNodeChanges(changes, nds))
        },
        [],
    );
```
This ensures that nodes with the specified ID will not be deleted.
